var/spack/repos/builtin/packages/mesa/package.py

```diff
@@ -69,23 +77,53 @@ class Mesa(AutotoolsPackage):
 
     # Variant dependencies
     depends_on('llvm@6:', when='+llvm')
+
     depends_on('libx11',  when='+glx')
-    depends_on('libxcb',  when='+glx')
+    depends_on('libxcb@1.9.3:', when='+glx')
     depends_on('libxext', when='+glx')
+
+    depends_on('libglvnd', when='+glvnd')
+
+    depends_on('libxshmfence@1.1:', when='+glvnd')
+
+    plus_dri = ('+egl', '+glvnd')
+    multi_depends('libdrm', when=plus_dri)
+    multi_depends('dri2proto', when=plus_dri)
+
+    depends_on('libxdamage@1.1:', when='+glvnd')
+    depends_on('damageproto', when='+glvnd')
+    depends_on('libxfixes', when='+glvnd')
+    depends_on('fixesproto', when='+glvnd')
+    depends_on('libxxf86vm', when='+glvnd')
+    depends_on('xf86vidmodeproto', when='+glvnd')
+
     depends_on('glproto@1.4.14:', when='+glx', type='build')
 
+    conflicts('+glvnd~glx~egl',
+              msg='+glvnd requires at least one of either +glx or +egl')
+
     # Prevent an unnecessary xcb-dri dependency
     patch('autotools-x11-nodri.patch')
+    patch('glvnd-install-egl-file-under-prefix.patch', when='+glvnd')
 
     def autoreconf(self, spec, prefix):
         which('autoreconf')('--force',  '--verbose', '--install')
 
     def configure_args(self):
         spec = self.spec
+
+        glvnd_enabled = '+glvnd' in spec
+        egl_enabled = '+egl' in spec
+        glx_enabled = '+glx' in spec
+        dri_enabled = egl_enabled or glvnd_enabled
+
+        enable_glvnd = ''.join((
+            '--', 'enable' if glvnd_enabled else 'disable', '-libglvnd'))
+
         args = [
             '--enable-shared',
             '--disable-static',
-            '--disable-libglvnd',
+            enable_glvnd,
```

Comment:
Add the `--{en,dis}able-dri` option here instead of mixing it with the glx and egl options down below